pkg/apis/v1beta1/nodepool.go

```diff
@@ -228,30 +248,36 @@ func (nl *NodePoolList) OrderByWeight() {
 	})
 }
 
-// MustGetAllowedDisruptions calls GetAllowedDisruptions and returns 0 if the error is not nil. This reduces the
-// amount of state that the disruption controller must reconcile, while allowing the GetAllowedDisruptions()
+// MustGetAllowedDisruptions calls GetAllowedDisruptionsByReason if the error is not nil. This reduces the
+// amount of state that the disruption controller must reconcile, while allowing the GetAllowedDisruptionsByReason()
 // to bubble up any errors in validation.
-func (in *NodePool) MustGetAllowedDisruptions(ctx context.Context, c clock.Clock, numNodes int) int {
-	val, err := in.GetAllowedDisruptions(ctx, c, numNodes)
+func (in *NodePool) MustGetAllowedDisruptions(ctx context.Context, c clock.Clock, numNodes int) map[DisruptionReason]int {
+	allowedDisruptions, err := in.GetAllowedDisruptionsByReason(ctx, c, numNodes)
 	if err != nil {
-		return 0
+		return map[DisruptionReason]int{}
 	}
-	return val
+	return allowedDisruptions
 }
 
-// GetAllowedDisruptions returns the minimum allowed disruptions across all disruption budgets for a given node pool.
-// This will return an error if there is a configuration error with any budget's node or schedule values.
-func (in *NodePool) GetAllowedDisruptions(ctx context.Context, c clock.Clock, numNodes int) (int, error) {
-	minVal := math.MaxInt32
-	var multiErr error
-	for i := range in.Spec.Disruption.Budgets {
-		val, err := in.Spec.Disruption.Budgets[i].GetAllowedDisruptions(c, numNodes)
+// GetAllowedDisruptionsByReason returns the minimum allowed disruptions across all disruption budgets, for all disruption methods for a given nodepool
+func (in *NodePool) GetAllowedDisruptionsByReason(ctx context.Context, c clock.Clock, numNodes int) (map[DisruptionReason]int, error) {
+	allowedDisruptions := map[DisruptionReason]int{}
+	for _, reason := range WellKnownDisruptionReasons {
+		allowedDisruptions[reason] = math.MaxInt32
+	}
+
+	for _, budget := range in.Spec.Disruption.Budgets {
+		val, err := budget.GetAllowedDisruptions(c, numNodes)
 		if err != nil {
-			multiErr = multierr.Append(multiErr, err)
+			return nil, err
```

Comment:
Can we return to `multierr` for this? I think it's more consistent with the rest of the repo and allows us to see if there are errors with more than the first failing disruption reason.

Reply:
Good catch